File: RndTacToe/RndTacToe.Lobby.Core/LobbyService.cs
using Microsoft.EntityFrameworkCore;
using RndTacToe.Lobby.DataAccess;
using RndTacToe.Models.LobbyDtos;
using RndTacToe.Models.LobbyEntities;

namespace RndTacToe.Lobby.Core
{
    public class LobbyService : ILobbyService
    {
        private readonly LobbyContext _context;

        public LobbyService(LobbyContext context)
        {
            _context = context;
        }

        public async Task<IEnumerable<LobbyDto>> GetLobbyAsync()
        {
            return await _context.Rooms.Select(x => new LobbyDto(x.GameId, x.PlayerName)).ToListAsync();
        }

        public async Task<LobbyDto?> GetLobbyDetail(string gameId)
        {
            var roomToFind = await _context.Rooms.FirstOrDefaultAsync(x => x.GameId == gameId);
            if (roomToFind != null)
            {
                return new LobbyDto(gameId, roomToFind.PlayerName);
            }
            return null;
        }

        public async Task<NewRoomResponse> NewGameAsync(NewRoomDto newRoom)
        {
            try
            {
                var lobbies = await GetLobbyAsync();
                if (lobbies.Count() >= 50)
                {
                    return new NewRoomResponse
                    {
                        IsCreated = false,
                        Reason = "Lobby is full, please wait to create another game"
                    };
                }
                var roomToInsert = new Room(newRoom.GameId, newRoom.PlayerName);
                await _context.Rooms.AddAsync(roomToInsert);
                await _context.SaveChangesAsync();
                return new NewRoomResponse()
                {
                    IsCreated = true
                };
            }
            catch (Exception ex)
            {

                return new NewRoomResponse
                {
                    IsCreated = false,
                    Reason = ex.Message
                };
            }
        }

        public async Task<bool> DeleteGameAsync(string gameId)
        {
            try
            {
                var toDelete = await _context.Rooms.SingleAsync(x => x.GameId == gameId);
                if (toDelete != null)
                {
                    _context.Rooms.Remove(toDelete);
                    await _context.SaveChangesAsync();
                }
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }
    }
}

File: server/services/lobby.service.ts
import { Room } from '../models/room.model';
import redis from '../utils/redis';

const getLobby = async (): Promise<Room[]> => {
  const lobby = await redis.get('lobby');
  return lobby ? (JSON.parse(lobby) as Room[]) : [];
};

const getLobbyDetail = async (gameId: string): Promise<Room | null> => {
  const lobby = await getLobby();
  return lobby.find((room) => room.gameId === gameId) ?? null;
};

const createNewRoom = async (room: Room): Promise<void> => {
  let lobby = await getLobby();

  if (!lobby?.length) {
    lobby = [];
  }

  if (lobby.find((x) => x.gameId === room.gameId)) {
    throw new Error('Session already exists');
  }

  if (lobby?.length > 50) {
    throw new Error('Lobby is full, please wait to create another game');
  }

  lobby.push(room);

  await redis.set('lobby', JSON.stringify(lobby));
};

export default {
  getLobby,
  getLobbyDetail,
  createNewRoom
};